diff --git a/apps/readrange/Makefile b/apps/readrange/Makefile
index 7076534a..c1dc140e 100644
--- a/apps/readrange/Makefile
+++ b/apps/readrange/Makefile
@@ -6,6 +6,7 @@ TARGET = bacrr
 BACNET_OBJECT_DIR = $(BACNET_SRC_DIR)/bacnet/basic/object
 SRC = main.c \
 	$(BACNET_OBJECT_DIR)/client/device-client.c \
+	$(BACNET_OBJECT_DIR)/trendlog.c \
 	$(BACNET_OBJECT_DIR)/netport.c
 
 # TARGET_EXT is defined in apps/Makefile as .exe or nothing
diff --git a/apps/readrange/main.c b/apps/readrange/main.c
index 5f003889..2761c4c0 100644
--- a/apps/readrange/main.c
+++ b/apps/readrange/main.c
@@ -194,10 +194,10 @@ static void print_help(char *filename)
         "you could send the following commands:\n");
     printf("%s 123 trend-log 2 log-buffer 1 1 10\n", filename);
     printf("%s 123 trend-log 2 log-buffer 2 1 10\n", filename);
-    printf("%s 123 trend-log 2 log-buffer 3 1/1/2014 00:00:01 10\n", filename);
+    printf("%s 123 trend-log 2 log-buffer 3 2014/1/1 00:00:01 10\n", filename);
     printf("%s 123 20 2 131 1 1 10\n", filename);
     printf("%s 123 20 2 131 2 1 10\n", filename);
-    printf("%s 123 20 2 131 3 1/1/2014 00:00:01 10\n", filename);
+    printf("%s 123 20 2 131 3 2014/1/1 00:00:01 10\n", filename);
 }
 
 int main(int argc, char *argv[])
@@ -261,7 +261,7 @@ int main(int argc, char *argv[])
         return 1;
     }
     if (Target_Object_Range_Type == 1) {
-        if (argc < 7) {
+        if (argc < 8) {
             print_usage(filename);
             return 0;
         }
@@ -271,7 +271,7 @@ int main(int argc, char *argv[])
         RR_Request.Range.RefIndex = Target_Object_Index;
         RR_Request.Count = Target_Object_Count;
     } else if (Target_Object_Range_Type == 2) {
-        if (argc < 7) {
+        if (argc < 8) {
             print_usage(filename);
             return 0;
         }
@@ -281,7 +281,7 @@ int main(int argc, char *argv[])
         RR_Request.Range.RefSeqNum = Target_Object_Index;
         RR_Request.Count = Target_Object_Count;
     } else if (Target_Object_Range_Type == 3) {
-        if (argc < 8) {
+        if (argc < 9) {
             print_usage(filename);
             return 0;
         }
diff --git a/src/bacnet/bacapp.c b/src/bacnet/bacapp.c
index d12bc84b..73ab9528 100644
--- a/src/bacnet/bacapp.c
+++ b/src/bacnet/bacapp.c
@@ -2536,6 +2536,64 @@ int bacapp_snprintf_value(
                 ret_val += slen;
                 break;
 #endif
+#if defined(BACAPP_DEVICE_OBJECT_PROPERTY_REFERENCE)
+            case BACNET_APPLICATION_TAG_DEVICE_OBJECT_PROPERTY_REFERENCE:
+                slen = snprintf(str, str_len, "%i ",
+                        value->type.Device_Object_Property_Reference.deviceIdentifier.instance);
+                if (str) {
+                    str += slen;
+                    if (str_len >= slen) {
+                        str_len -= slen;
+                    } else {
+                        str_len = 0;
+                    }
+                }
+                ret_val += slen;
+                if (value->type.Device_Object_Property_Reference.objectIdentifier.type <= BACNET_OBJECT_TYPE_LAST) {
+                    slen = snprintf(str, str_len, "%s ",
+                        bactext_object_type_name(value->type.Device_Object_Property_Reference.objectIdentifier.type));
+                } else {
+                    slen = snprintf(str, str_len, "%u ",
+                        (unsigned)value->type.Device_Object_Property_Reference.objectIdentifier.type);
+                }
+                if (str) {
+                    str += slen;
+                    if (str_len >= slen) {
+                        str_len -= slen;
+                    } else {
+                        str_len = 0;
+                    }
+                }
+                ret_val += slen;
+                slen = snprintf(str, str_len, "%i ",
+                        value->type.Device_Object_Property_Reference.objectIdentifier.instance);
+                if (str) {
+                    str += slen;
+                    if (str_len >= slen) {
+                        str_len -= slen;
+                    } else {
+                        str_len = 0;
+                    }
+                }
+                ret_val += slen;
+                if (value->type.Device_Object_Property_Reference.objectIdentifier.type <= MAX_BACNET_PROPERTY_ID) {
+                    slen = snprintf(str, str_len, "%s",
+                        bactext_property_name(value->type.Device_Object_Property_Reference.propertyIdentifier));
+                } else {
+                    slen = snprintf(str, str_len, "%u",
+                        (unsigned)value->type.Device_Object_Property_Reference.propertyIdentifier);
+                }
+                if (str) {
+                    str += slen;
+                    if (str_len >= slen) {
+                        str_len -= slen;
+                    } else {
+                        str_len = 0;
+                    }
+                }
+                ret_val += slen;
+                break;
+#endif
 #if defined(BACAPP_LIGHTING_COMMAND)
             case BACNET_APPLICATION_TAG_LIGHTING_COMMAND:
                 ret_val = lighting_command_to_ascii(
diff --git a/src/bacnet/bacapp.h b/src/bacnet/bacapp.h
index 57381740..8403c162 100644
--- a/src/bacnet/bacapp.h
+++ b/src/bacnet/bacapp.h
@@ -173,6 +173,15 @@ typedef struct BACnet_Object_Property_Value {
     BACNET_APPLICATION_DATA_VALUE *value;
 } BACNET_OBJECT_PROPERTY_VALUE;
 
+struct BACnet_Trendlog_Record;
+typedef struct BACnet_Trendlog_Record {
+  BACNET_DATE_TIME timestamp;
+  BACNET_APPLICATION_DATA_VALUE value;
+  BACNET_BIT_STRING status;
+  struct BACnet_Trendlog_Record *next;
+} BACNET_TRENDLOG_RECORD;
+
+
 #ifdef __cplusplus
 extern "C" {
 #endif /* __cplusplus */
diff --git a/src/bacnet/basic/object/trendlog.c b/src/bacnet/basic/object/trendlog.c
index b8da5b19..3e3a2c0e 100644
--- a/src/bacnet/basic/object/trendlog.c
+++ b/src/bacnet/basic/object/trendlog.c
@@ -26,6 +26,7 @@
 #include <stdbool.h>
 #include <stdint.h>
 #include <string.h> /* for memmove */
+#include <stdlib.h> /* calloc */
 #include "bacnet/bacdef.h"
 #include "bacnet/bacdcode.h"
 #include "bacnet/bacenum.h"
@@ -160,7 +161,7 @@ void Trend_Log_Init(void)
     static bool initialized = false;
     int iLog;
     int iEntry;
-    BACNET_DATE_TIME bdatetime = { 0 };
+    BACNET_DATE_TIME bdatetime = { { 0 }, { 0 } };
     bacnet_time_t tClock;
     uint8_t month;
 
@@ -1513,6 +1514,122 @@ int TL_encode_entry(uint8_t *apdu, int iLog, int iEntry)
     return (iLen);
 }
 
+int rr_decode_trendlog_entries(
+    uint8_t *apdu, int apdu_len, BACNET_TRENDLOG_RECORD *rec)
+{
+    int len;
+    int status = -1;
+    bool tag0 = false;
+    bool tag1 = false;
+    bool tag2 = false;
+    uint8_t tag_number = 0;
+    uint32_t len_value_type = 0;
+    rec->next = NULL;
+    while (apdu_len > 0) {
+        if (IS_CONTEXT_SPECIFIC(apdu[0]) &&
+            decode_is_opening_tag_number(apdu, 0) &&
+            !tag0) {
+            tag0 = true;
+
+            len = bacapp_decode_context_datetime(apdu, 0, &rec->timestamp);
+            if (len <= 0) {
+                return -1;
+            }
+            apdu += len;
+            apdu_len -= len;
+            status = 1;
+        } else if (IS_CONTEXT_SPECIFIC(apdu[0]) &&
+            decode_is_opening_tag_number(apdu, 1) &&
+            !tag1) {
+            tag1 = true;
+            // skip the opening tag
+            apdu++;
+            apdu_len--;
+
+            // decode the next context tag which has th value type
+            len =
+                decode_tag_number_and_value(apdu, &tag_number, &len_value_type);
+            if (len <= 0) {
+                return -1;
+            }
+            status = 1;
+            switch (tag_number) {
+                case TL_TYPE_BOOL:
+                    rec->value.tag = BACNET_APPLICATION_TAG_BOOLEAN;
+                    len = decode_context_boolean2(
+                        apdu, tag_number, &rec->value.type.Boolean);
+                    break;
+                case TL_TYPE_REAL:
+                    rec->value.tag = BACNET_APPLICATION_TAG_REAL;
+                    len = decode_context_real(
+                        apdu, tag_number, &rec->value.type.Real);
+                    break;
+                case TL_TYPE_ENUM:
+                    rec->value.tag = BACNET_APPLICATION_TAG_ENUMERATED;
+                    len = decode_context_enumerated(
+                        apdu, tag_number, &rec->value.type.Enumerated);
+                    break;
+                case TL_TYPE_UNSIGN:
+                    rec->value.tag = BACNET_APPLICATION_TAG_UNSIGNED_INT;
+                    len = decode_context_unsigned(
+                        apdu, tag_number, &rec->value.type.Unsigned_Int);
+                    break;
+                case TL_TYPE_SIGN:
+                    rec->value.tag = BACNET_APPLICATION_TAG_SIGNED_INT;
+                    len = decode_context_signed(
+                        apdu, tag_number, &rec->value.type.Signed_Int);
+                    break;
+                case TL_TYPE_BITS:
+                    rec->value.tag = BACNET_APPLICATION_TAG_BIT_STRING;
+                    len = decode_context_bitstring(
+                        apdu, tag_number, &rec->value.type.Bit_String);
+                    break;
+                case TL_TYPE_NULL:
+                    rec->value.tag = BACNET_APPLICATION_TAG_NULL;
+                    break;
+                default:
+                    // skip over the value if we don't suppord decoding it
+                    len += len_value_type;
+            }
+            if (len <= 0) {
+                break;
+            }
+            apdu += len;
+            apdu_len -= len;
+
+            // skip over the closing tag [1]
+            if (IS_CONTEXT_SPECIFIC(apdu[0]) &&
+                decode_is_closing_tag_number(apdu, 1)) {
+                apdu++;
+                apdu_len--;
+            } else {
+                return -1;
+            }
+        } else if (IS_CONTEXT_SPECIFIC(apdu[0]) &&
+            !tag2) {
+            tag2 = true;
+            // context tag 2 is a status bitstring.
+            // we don't do anything with this other than decode it.
+            len = decode_context_bitstring(&apdu[0], 2, &rec->status);
+            if (len > 0) {
+                status = 1;
+                apdu += len;
+                apdu_len -= len;
+            }
+        } else {
+            if (apdu_len > 0) {
+                tag0 = false;
+                tag1 = false;
+                tag2 = false;
+                rec->next = calloc(sizeof(BACNET_TRENDLOG_RECORD), 1);
+                rec = rec->next;
+                rec->next = NULL;
+            }
+        }
+    }
+    return status;
+}
+
 static int local_read_property(uint8_t *value,
     uint8_t *status,
     BACNET_DEVICE_OBJECT_PROPERTY_REFERENCE *Source,
diff --git a/src/bacnet/basic/object/trendlog.h b/src/bacnet/basic/object/trendlog.h
index 087ff26e..f46c964a 100644
--- a/src/bacnet/basic/object/trendlog.h
+++ b/src/bacnet/basic/object/trendlog.h
@@ -213,6 +213,12 @@ extern "C" {
         uint8_t * apdu,
         BACNET_READ_RANGE_DATA * pRequest);
 
+    BACNET_STACK_EXPORT
+    int rr_decode_trendlog_entries(
+        uint8_t *apdu,
+	int apdu_len,
+	BACNET_TRENDLOG_RECORD *pRec);
+ 
     BACNET_STACK_EXPORT
     void trend_log_timer(
         uint16_t uSeconds);
diff --git a/src/bacnet/basic/service/h_rr_a.c b/src/bacnet/basic/service/h_rr_a.c
index a6a2f485..b5682367 100644
--- a/src/bacnet/basic/service/h_rr_a.c
+++ b/src/bacnet/basic/service/h_rr_a.c
@@ -36,6 +36,7 @@
 #include "bacnet/basic/object/device.h"
 #include "bacnet/basic/services.h"
 #include "bacnet/basic/tsm/tsm.h"
+#include "bacnet/basic/object/trendlog.h"
 
 /** @file h_rr_a.c  Handles Read Range Acknowledgments. */
 
@@ -43,59 +44,48 @@
 static void PrintReadRangeData(BACNET_READ_RANGE_DATA *data)
 {
 #ifdef BACAPP_PRINT_ENABLED
-    BACNET_OBJECT_PROPERTY_VALUE object_value; /* for bacapp printing */
-#endif
-    BACNET_APPLICATION_DATA_VALUE value; /* for decode value data */
-    int len = 0;
-    uint8_t *application_data;
-    int application_data_len;
-    bool first_value = true;
-#if PRINT_ENABLED
-    bool print_brace = false;
+    BACNET_OBJECT_PROPERTY_VALUE object_value;
 #endif
+    BACNET_APPLICATION_DATA_VALUE value;
+    BACNET_TRENDLOG_RECORD entry;
+    BACNET_TRENDLOG_RECORD *p;
+    int status = 0;
 
     if (data) {
-        application_data = data->application_data;
-        application_data_len = data->application_data_len;
+#ifdef BACAPP_PRINT_ENABLED
+        object_value.object_type = data->object_type;
+        object_value.object_instance = data->object_instance;
+        object_value.object_property = data->object_property;
+        object_value.array_index = data->array_index;
+#endif
+
         /* FIXME: what if application_data_len is bigger than 255? */
         /* value? need to loop until all of the len is gone... */
-        for (;;) {
-            len = bacapp_decode_application_data(
-                application_data, (uint8_t)application_data_len, &value);
-            if (first_value && (len < application_data_len)) {
-                first_value = false;
-#if PRINT_ENABLED
-                fprintf(stdout, "{");
-                print_brace = true;
-#endif
-            }
+
+        status = rr_decode_trendlog_entries(
+            data->application_data, data->application_data_len, &entry);
 #ifdef BACAPP_PRINT_ENABLED
-            object_value.object_type = data->object_type;
-            object_value.object_instance = data->object_instance;
-            object_value.object_property = data->object_property;
-            object_value.array_index = data->array_index;
+        if (status < 1) {
+            return;
+        }
+        printf("{\"list\": [\n");
+        for (p = &entry; p != NULL; p = p->next) {
+            printf(" [\"");
             object_value.value = &value;
+            value.tag = BACNET_APPLICATION_TAG_TIMESTAMP;
+            value.type.Time_Stamp.value.dateTime = p->timestamp;
             bacapp_print_value(stdout, &object_value);
-#endif
-            if (len > 0) {
-                if (len < application_data_len) {
-                    application_data += len;
-                    application_data_len -= len;
-                    /* there's more! */
-#if PRINT_ENABLED
-                    fprintf(stdout, ",");
-#endif
-                } else {
-                    break;
-                }
-            } else {
-                break;
-            }
+            printf("\",\"");
+
+            object_value.value = &p->value;
+            bacapp_print_value(stdout, &object_value);
+
+            if (p->next)
+                printf("\"],\n");
+            else
+                printf("\"]\n");
         }
-#if PRINT_ENABLED
-        if (print_brace)
-            fprintf(stdout, "}");
-        fprintf(stdout, "\r\n");
+        printf("]}\n");
 #endif
     }
 }
