diff --git a/src/bacnet/bacapp.c b/src/bacnet/bacapp.c
index 7b34acda..b6c7eda1 100644
--- a/src/bacnet/bacapp.c
+++ b/src/bacnet/bacapp.c
@@ -1907,16 +1907,6 @@ int bacapp_snprintf_value(
             case BACNET_APPLICATION_TAG_CHARACTER_STRING:
                 len = characterstring_length(&value->type.Character_String);
                 char_str = characterstring_value(&value->type.Character_String);
-                slen = snprintf(str, str_len, "\"");
-                if (str) {
-                    str += slen;
-                    if (str_len >= slen) {
-                        str_len -= slen;
-                    } else {
-                        str_len = 0;
-                    }
-                }
-                ret_val += slen;
 #ifdef __STDC_ISO_10646__
                 if (characterstring_encoding(&value->type.Character_String) ==
                     CHARACTER_UTF8) {
@@ -1956,11 +1946,7 @@ int bacapp_snprintf_value(
 #endif
                 {
                     for (i = 0; i < len; i++) {
-                        if (isprint(*((unsigned char *)char_str))) {
-                            slen = snprintf(str, str_len, "%c", *char_str);
-                        } else {
-                            slen = snprintf(str, str_len, "%c", '.');
-                        }
+                        slen = snprintf(str, str_len, "%c", *char_str);
                         if (str) {
                             str += slen;
                             if (str_len >= slen) {
@@ -1973,8 +1959,6 @@ int bacapp_snprintf_value(
                         char_str++;
                     }
                 }
-                slen = snprintf(str, str_len, "\"");
-                ret_val += slen;
                 break;
 #endif
 #if defined(BACAPP_BIT_STRING)
@@ -2243,22 +2227,9 @@ int bacapp_snprintf_value(
                     name = &value->type.Host_Address.host.name;
                     len = characterstring_length(name);
                     char_str = characterstring_value(name);
-                    slen = snprintf(str, str_len, "\"");
-                    if (str) {
-                        str += slen;
-                        if (str_len >= slen) {
-                            str_len -= slen;
-                        } else {
-                            str_len = 0;
-                        }
-                    }
                     ret_val += slen;
                     for (i = 0; i < len; i++) {
-                        if (isprint(*((unsigned char *)char_str))) {
-                            slen = snprintf(str, str_len, "%c", *char_str);
-                        } else {
-                            slen = snprintf(str, str_len, "%c", '.');
-                        }
+                        slen = snprintf(str, str_len, "%c", *char_str);
                         char_str++;
                         if (str) {
                             str += slen;
@@ -2270,8 +2241,6 @@ int bacapp_snprintf_value(
                         }
                         ret_val += slen;
                     }
-                    slen = snprintf(str, str_len, "\"");
-                    ret_val += slen;
                 }
                 break;
 #endif
