diff --git a/apps/whois/main.c b/apps/whois/main.c
index 20279a9a1..6c503f04a 100644
--- a/apps/whois/main.c
+++ b/apps/whois/main.c
@@ -43,6 +43,11 @@ static bool Error_Detected = false;
 /* debug info printing */
 static bool BACnet_Debug_Enabled;
 
+#if defined(BACDL_BIP)
+/* If set, use this as the source port. */
+static uint16_t My_BIP_Port = 0;
+#endif
+
 #define BAC_ADDRESS_MULT 1
 
 struct address_entry {
@@ -414,6 +419,10 @@ int main(int argc, char *argv[])
             if (++argi < argc) {
                 delay_milliseconds = strtol(argv[argi], NULL, 0);
             }
+        } else if (strcmp(argv[argi], "--sport") == 0) {
+            if (++argi < argc) {
+                My_BIP_Port = (uint16_t)strtol(argv[argi], NULL, 0);
+            }
         } else {
             if (target_args == 0) {
                 Target_Object_Instance_Min = Target_Object_Instance_Max =
@@ -474,6 +483,15 @@ int main(int argc, char *argv[])
     }
     /* setup my info */
     Device_Set_Object_Instance_Number(BACNET_MAX_INSTANCE);
+#if defined(BACDL_BIP)
+    /* For BACnet/IP, we might have set a different port for "me", so
+     * (eg) we could talk to a BACnet/IP device on our same interface.
+     * My_BIP_Port will be non-zero in this case.
+     */
+    if (My_BIP_Port > 0) {
+        bip_set_port(My_BIP_Port);
+    }
+#endif
     init_service_handlers();
     address_init();
     dlenv_init();
@@ -483,6 +501,17 @@ int main(int argc, char *argv[])
     }
     mstimer_set(&apdu_timer, timeout_milliseconds);
     mstimer_set(&datalink_timer, 1000);
+#if defined(BACDL_BIP)
+    if (My_BIP_Port > 0) {
+        char *pEnv = NULL; pEnv = getenv("BACNET_IP_PORT");
+        if (pEnv) {
+            bip_set_port((uint16_t)strtol(pEnv, NULL, 0));
+        } else {
+            /* Set back to std BACnet/IP port */
+            bip_set_port(0xBAC0);
+        }
+    }
+#endif
     /* send the request */
     Send_WhoIs_To_Network(
         &dest, Target_Object_Instance_Min, Target_Object_Instance_Max);
